src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -13,36 +13,64 @@ namespace System.Net.Http
     /// <summary>
     /// DiagnosticHandler notifies DiagnosticSource subscribers about outgoing Http requests
     /// </summary>
-    internal sealed class DiagnosticsHandler : DelegatingHandler
+    internal sealed class DiagnosticsHandler : HttpMessageHandlerStage
     {
         private static readonly DiagnosticListener s_diagnosticListener =
                 new DiagnosticListener(DiagnosticsHandlerLoggingStrings.DiagnosticListenerName);
 
-        /// <summary>
-        /// DiagnosticHandler constructor
-        /// </summary>
-        /// <param name="innerHandler">Inner handler: Windows or Unix implementation of HttpMessageHandler.
-        /// Note that DiagnosticHandler is the latest in the pipeline </param>
-        public DiagnosticsHandler(HttpMessageHandler innerHandler) : base(innerHandler)
+        private readonly HttpMessageHandler _innerHandler;
+        private readonly TextMapPropagator _propagator;
+        private readonly HeaderDescriptor[]? _propagatorFields;
+
+        public DiagnosticsHandler(HttpMessageHandler innerHandler, TextMapPropagator? propagator = null, bool autoRedirect = false)
         {
+            Debug.Assert(IsGloballyEnabled());
+
+            _innerHandler = innerHandler;
+            _propagator = propagator ?? TextMapPropagator.Current;
+
+            if (autoRedirect)
+            {
+                // Prepare HeaderDescriptors for fields we need to clear when following redirects
+                // If a PassThroughPropagator is used, we don't have to clear the headers as they will always be the same
+                if (_propagator.Fields is IReadOnlyCollection<string> fields &&
+                    fields.Count > 0 &&
+                    !ReferenceEquals(_propagator, TextMapPropagator.CreatePassThroughPropagator()))
+                {
+                    var fieldDescriptors = new List<HeaderDescriptor>(fields.Count);
+                    foreach (string field in fields)
+                    {
+                        if (field is not null && HeaderDescriptor.TryGet(field, out HeaderDescriptor descriptor))
+                        {
+                            fieldDescriptors.Add(descriptor);
+                        }
+                    }
+                    _propagatorFields = fieldDescriptors.ToArray();
+                }
+            }
         }
 
-        internal static bool IsEnabled()
+        private static bool IsEnabled()
```

Comment:
>IsEnabled [](http://example.com/codeflow?start=28&length=9)

nit: we usaully use => syntax in the one liner methods. I am not sure if you apply this style here.

Reply:
We usually do the same. I left it as-is here to keep the comment in scope (and the `ActivitySource` change will remove this method anyway).